pkg/controller/kubedirectorcluster/cluster.go

```diff
@@ -57,6 +57,65 @@ func (r *ReconcileKubeDirectorCluster) syncCluster(
 		cr.Status = &kdv1.KubeDirectorClusterStatus{}
 		cr.Status.Roles = make([]kdv1.RoleStatus, 0)
 	}
+
+	// Check, if some statefulset is in the middle of upgrade/rollback process.
+	// If so, the current cluster state still can be in `configured` state, change it to `updating`
+	upgradeIsActive := false
+	for _, rs := range cr.Status.Roles {
+		if rs.RoleUpgradeStatus == kdv1.RoleRollingBack || rs.RoleUpgradeStatus == kdv1.RoleUpgrading {
+			upgradeIsActive = true
+			cr.Status.State = string(clusterUpdating)
+			break
+		}
+	}
+
+	// If all statefulsets completed their upgrade/rollback processes
+	// erase the cluster upgradeInfo object
+	if cr.Status.UpgradeInfo != nil && !upgradeIsActive {
+		// If there wasn't rollback
+		if cr.Spec.AppID != cr.Status.UpgradeInfo.PrevApp {
+			// Remove the cluster bind with the previous app
+			shared.RemoveClusterAppReference(
+				cr.Namespace,
+				cr.Name,
+				*cr.Spec.AppCatalog,
+				cr.Status.UpgradeInfo.PrevApp,
+			)
+			// And bind the cluster with the new app
+			shared.EnsureClusterAppReference(
+				cr.Namespace,
+				cr.Name,
+				*cr.Spec.AppCatalog,
+				cr.Spec.AppID,
+			)
+		}
+		cr.Status.UpgradeInfo = nil
+	}
+
+	upgradeInfo := (*cr).Status.UpgradeInfo
+
+	// KD cluter is created recently, init the Status.AppID field
+	if cr.Status.AppID == nil {
+		cr.Status.AppID = &cr.Spec.AppID
+	} else
+	// Spec.AppID was changed, create the upgradeInfo object and store there the current AppID
+	// Also, change cluster state to `updating`
+	if cr.Spec.AppID != *cr.Status.AppID && upgradeInfo == nil {
+		cr.Status.UpgradeInfo = &kdv1.UpgradeInfo{
+			IsRollingBack: false,
+			PrevApp:       *cr.Status.AppID,
+		}
+		cr.Status.AppID = &cr.Spec.AppID
+		(*cr).Status.State = string(clusterUpdating)
+	} else
+	// If KD cluster upgradeInfo object exists and is not in RollingBack status,
+	// but the current cluster AppID is the same as in the upgradeInfo
+	// it means that rollback is required
+	if upgradeInfo != nil && !upgradeInfo.IsRollingBack && cr.Spec.AppID == upgradeInfo.PrevApp {
+		upgradeInfo.IsRollingBack = true
+		cr.Status.AppID = &upgradeInfo.PrevApp
```

Comment:
I think we also should swap prevApp back to the app we are rolling-back-from. That will come in useful elsewhere I think. So this code could do:
```go
upgradeInfo.PrevApp = *cr.Status.AppID
cr.Status.AppID = &cr.Spec.AppID
```

Reply:
Done